src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/FrameConnection.cs

```diff
@@ -454,5 +460,68 @@ public void StopTimingWrite()
                 _writeTimingWrites--;
             }
         }
+
+        private IDisposable BeginConnectionScope()
+        {
+            if (Log.IsEnabled(LogLevel.Critical))
+            {
+                return Log.BeginScope(new ConnectionScope(ConnectionId));
+            }
+
+            return null;
+        }
+
+        private class ConnectionScope : IReadOnlyList<KeyValuePair<string, object>>
```

Comment:
That's also true for all the dozens of what used to be private classes we moved to their own files in Kestrel. FrameConnection is already pretty big.